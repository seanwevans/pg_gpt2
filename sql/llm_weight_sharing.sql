SET client_min_messages = warning;
TRUNCATE llm_param,
         llm_param_share,
         llm_tensor,
         llm_tensor_rt,
         llm_tensor_map,
         llm_tape;

INSERT INTO llm_param(model, name, token_id, data, grad, m, v, step)
VALUES
    ('model_a', 'shared', 0, decode('0000803f', 'hex'), NULL, NULL, NULL, 0),
    ('model_b', 'unique', 0, decode('00000000', 'hex'), NULL, NULL, NULL, 0);

SELECT llm_share_param('model_a', 'shared', 0, 'model_b', 'shared', 0);

SELECT count(*) AS model_b_shared_rows
FROM llm_param
WHERE model = 'model_b'
  AND name = 'shared';

SELECT encode(data, 'hex') AS resolved_shared
FROM llm_param_resolved
WHERE model = 'model_b'
  AND name = 'shared';

SELECT source_model, source_name, target_model, target_name
FROM llm_param_share
ORDER BY 1, 2, 3, 4;

SELECT llm_materialize_params('model_b');

UPDATE llm_tensor_rt
SET grad = decode('00000040', 'hex')
WHERE id IN (
    SELECT tensor_id
    FROM llm_tensor_map
    WHERE model = 'model_b'
      AND name = 'shared'
      AND token_id = 0
);

SELECT llm_accumulate_grads('model_b');

SELECT encode(grad, 'hex') AS shared_grad
FROM llm_param
WHERE model = 'model_a'
  AND name = 'shared';

SELECT llm_unshare_param('model_b', 'shared', 0);

SELECT encode(data, 'hex') AS unshared_data, step
FROM llm_param
WHERE model = 'model_b'
  AND name = 'shared';

SELECT count(*) AS remaining_shares
FROM llm_param_share
WHERE target_model = 'model_b';
