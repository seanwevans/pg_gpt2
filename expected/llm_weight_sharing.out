SET
TRUNCATE TABLE
INSERT 0 2
 llm_share_param 
-----------------
 
(1 row)

 model_b_shared_rows 
----------------------
                    0
(1 row)

 resolved_shared 
-----------------
 0000803f
(1 row)

 source_model | source_name | target_model | target_name 
--------------+-------------+--------------+-------------
 model_a      | shared      | model_b      | shared
(1 row)

 llm_materialize_params 
------------------------
 
(1 row)

UPDATE 1
 llm_accumulate_grads 
----------------------
 
(1 row)

 shared_grad 
-------------
 00000040
(1 row)

 llm_unshare_param 
-------------------
 
(1 row)

 unshared_data | step 
---------------+------
 0000803f      |    0
(1 row)

 remaining_shares 
------------------
                0
(1 row)
